tracking/translations_parser/parser.py

```diff
@@ -128,16 +138,17 @@ def parse_validation_log(
         """Parses a validation entry on multiple lines."""
         if ("valid",) not in headers or not (match := VALIDATION_RE.match(text)):
             return None
-        epoch, up, key, val = match.groups()
+        results = match.groupdict()
         # Replace items keys to match ValidationEpoch dataclass
-        key = key.replace("-", "_")
-        # Transform values to match output types
-        epoch, up = int(epoch), int(up)
-        val = ValidationEpoch.__annotations__[key](val)
+        key = results["key"].replace("-", "_")
+        epoch, up = int(results["ep"]), int(results["up"])
         entry = self._validation_entries[(epoch, up)]
-        entry[key] = val
+        # Transform values to match output types
+        entry[key] = ValidationEpoch.__annotations__[key](results["value"])
+        if results["stalled"] is not None:
+            entry[f"{key}_stalled"] = float(results["stalled"])
         # Build a validation epoch from multiple lines
-        expected_keys = set(ValidationEpoch.__annotations__.keys()) - {"epoch", "up", "perplexity"}
+        expected_keys = {"ce_mean_words", "chrf", "bleu_detok"}
```

Comment:
Just to double check: those metrics are specified here and can be changed to any metrics supported by Marian `--valid-metrics` [argument](https://marian-nmt.github.io/docs/cmd/marian/):
https://github.com/mozilla/firefox-translations-training/blob/fac7511f0f84af352c90502132a162a52009cee3/pipeline/train/train.sh#L41

Does this code support arbitrary metrics?

Reply:
Parser's code is not generic, as it uses a regex to match lines with validation data.
However I can keep the original code, so updating new metrics would be simpler (i.e. updating the regex and the ValidationEpoch dataclass). I'll add a bit of documentation about this.

Reply:
Ok, it's fine. I don't think we'll want to change this soon.